### PR TITLE
Upgrading SpanHelpers with Vector512

### PR DESCRIPTION
**Pull request by DeepakRajendrakumaran**
### What this PR does:
1. This PR upgrades `Span` implementation by enabling acceleration using AVX512 instructions when possible. This is achieved using. This is achieved by adding `Vector512 `paths in relevant `SpanHelper `implementations.
2. It also modified `Vector512.IsHardwareAccelerated` implementation so that it now returns `TRUE` only on targets where there is no throttling for AVX512(this is based on discussion with Tanner)

### Performance testing for the PR
Ran Microbenchmark( _--filter "System.Memory.*"_) and compared with main branch : This was with following thresholds for ResultComparator on ICX( _--threshold 5% --noise 1ns_). I have re-run the ones showing regression locally and DO NOT see any which are consistently slower
image

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs [hunk 1]
```diff
@@ -63,7 +63,68 @@ public static int IndexOf(ref byte searchSpace, int searchSpaceLength, ref byte
             // Based on http://0x80.pl/articles/simd-strfind.html#algorithm-1-generic-simd "Algorithm 1: Generic SIMD" by Wojciech Mula
             // Some details about the implementation can also be found in https://github.com/dotnet/runtime/pull/63285
         SEARCH_TWO_BYTES:
-            if (Vector256.IsHardwareAccelerated && searchSpaceMinusValueTailLength - Vector256<byte>.Count >= 0)
+            if (Vector512.IsHardwareAccelerated && searchSpaceMinusValueTailLength - Vector512<byte>.Count >= 0)
+            {
+                // Find the last unique (which is not equal to ch1) byte
+                // the algorithm is fine if both are equal, just a little bit less efficient
+                byte ch2Val = Unsafe.Add(ref value, valueTailLength);
+                nint ch1ch2Distance = valueTailLength;
```
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:71]: `valueTailLength` is of type `int`, so to avoid the sign-extending move use
```suggestion
                nint ch1ch2Distance = (nint)(uint)valueTailLength;
```
See [codegen](https://godbolt.org/z/qnbMGqncv)

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs [hunk 2]
```diff
@@ -63,7 +63,68 @@ public static int IndexOf(ref byte searchSpace, int searchSpaceLength, ref byte
...
+                    Vector512<byte> cmpCh1 = Vector512.Equals(ch1, Vector512.LoadUnsafe(ref searchSpace, (nuint)offset));
+                    Vector512<byte> cmpAnd = (cmpCh1 & cmpCh2).AsByte();
+
+                    // Early out: cmpAnd is all zeros
+                    if (cmpAnd != Vector512<byte>.Zero)
+                    {
+                        goto CANDIDATE_FOUND;
+                    }
+
+                LOOP_FOOTER:
```
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:97]: Codegen-wise: is it good to have these labels in the loop?

**Review by gfoidl — COMMENTED** [review]
Had just a quick view -- same applies to the `char` variant.

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs (hunk 2 above)
Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:97]: It is sort of hand-written PGO, less relevant with Dynamic PGO being enabled by default, but we have other targets.

**Comment by DeepakRajendrakumaran** [comment]
> Would be nice to see benchmarks for small inputs (which are more commonly used in these perf-sensitive primitives), maybe it even makes sense to wrap avx-512 path into a call?..

Yes. That's part of the plan. Currently ran it only for IndexOf() for Byte. Will add for rest soon

image

File: src/coreclr/jit/gentree.cpp [hunk 3]
```diff
@@ -23591,6 +23591,19 @@ GenTree* Compiler::gtNewSimdShuffleNode(
...
+            {
+                vecCns.u8[i] = (uint8_t)(vecCns.u8[i * elementSize] / elementSize);
+            }
+
+            op2                        = gtNewVconNode(type);
+            op2->AsVecCon()->gtSimdVal = vecCns;
+
+            // swap the operands to match the encoding requirements
+            retNode = gtNewSimdHWIntrinsicNode(type, op2, op1, NI_AVX512VBMI_PermuteVar64x8, simdBaseJitType, simdSize);
+        }
```
Comment (tannergooding) [inline src/coreclr/jit/gentree.cpp:23606]: It would be nice to extract this acceleration into its own PR and also cover Vector256 at the same time, if possible.

File: src/coreclr/jit/hwintrinsicxarch.cpp [hunk 4]
```diff
@@ -2361,7 +2361,7 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
             }
             else if (simdSize == 64)
             {
-                if (varTypeIsByte(simdBaseType))
+                if (varTypeIsByte(simdBaseType) && (!compExactlyDependsOn(InstructionSet_AVX512VBMI)))
```
Comment (tannergooding) [inline src/coreclr/jit/hwintrinsicxarch.cpp:2364]: nit: Unnecessary parentheses:
```suggestion
                if (varTypeIsByte(simdBaseType) && !compExactlyDependsOn(InstructionSet_AVX512VBMI))
```

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs [hunk 5]
```diff
@@ -63,7 +63,68 @@ public static int IndexOf(ref byte searchSpace, int searchSpaceLength, ref byte
             // Based on http://0x80.pl/articles/simd-strfind.html#algorithm-1-generic-simd "Algorithm 1: Generic SIMD" by Wojciech Mula
             // Some details about the implementation can also be found in https://github.com/dotnet/runtime/pull/63285
         SEARCH_TWO_BYTES:
-            if (Vector256.IsHardwareAccelerated && searchSpaceMinusValueTailLength - Vector256<byte>.Count >= 0)
+            if (Vector512.IsHardwareAccelerated && searchSpaceMinusValueTailLength - Vector512<byte>.Count >= 0)
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:66]: Given the current handling we have for first gen AVX-512 hardware in the JIT, we probably want to have an internal helper and make this check something like:
```suggestion
if (Vector512.IsHardwareAccelerated && !X86Base.Vector512Throttling && searchSpaceMinusValueTailLength - Vector512<byte>.Count >= 0)
```

The `X86Base.Vector512Throttling` check can just be an `internal static bool Vector512Throttling` intrinsic that returns the underlying `JitFlags`.

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs [hunk 6]
```diff
@@ -63,7 +63,68 @@ public static int IndexOf(ref byte searchSpace, int searchSpaceLength, ref byte
...
+            {
+                // Find the last unique (which is not equal to ch1) byte
+                // the algorithm is fine if both are equal, just a little bit less efficient
+                byte ch2Val = Unsafe.Add(ref value, valueTailLength);
+                nint ch1ch2Distance = (nint)(uint)valueTailLength;
+                while (ch2Val == value && ch1ch2Distance > 1)
+                    ch2Val = Unsafe.Add(ref value, --ch1ch2Distance);
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:73]: Nothing for you to do here, just calling out that this kind of duplication between `Vector512/256/128` is another reason why having an `ISimdVector` would be nice: https://github.com/dotnet/runtime/pull/76423

It might be feasible for us to define/expose for internal use only for the time being. Interested in @stephentoub's thoughts on it.

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs [hunk 7]
```diff
@@ -1109,6 +1401,10 @@ private static Vector128<byte> LoadVector128(ref byte start, nuint offset)
         private static Vector256<byte> LoadVector256(ref byte start, nuint offset)
             => Unsafe.ReadUnaligned<Vector256<byte>>(ref Unsafe.AddByteOffset(ref start, offset));
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Vector512<byte> LoadVector512(ref byte start, nuint offset)
+            => Unsafe.ReadUnaligned<Vector512<byte>>(ref Unsafe.AddByteOffset(ref start, offset));
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:1406]: These helpers shouldn't be needed anymore, we can just use `Vector512.LoadUnsafe(ref start, offset)` now.

**Comment by tannergooding** [comment]
The changes overall LGTM, there's just quite a lot of essentially duplicated code that we can't currently share between Vector512/256/128. I think we want to also consider how to avoid this negatively impacting 1st gen AVX-512 hardware and so I proposed an internal only property to hang off `X86Base`. We could consider exposing that more broadly as well, but that's a different discussion.

File: src/coreclr/jit/gentree.cpp (hunk 3 above)
Comment (DeepakRajendrakumaran) [reply src/coreclr/jit/gentree.cpp:23606]: Moved to separate PR

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs (hunk 5 above)
Comment (DeepakRajendrakumaran) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:66]: Does this make sense more as a flag/variable similar to 'IsSupported' or as an function to be inlined. I'm leaning more the former

Comment (DeepakRajendrakumaran) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:66]: Would I need to hide a hypothetical `X86Base.Vector512Throttling` check inside a `X86Base.IsSupported`or `CompatDependsOn `?

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:66]: I don't think so since its just returning a `bool` and and `X86Base` is considered part of the baseline ISA set, so isn't tracked.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:66]: However, we could put it on `Vector512` instead just to be safe. That would make it `Vector512.IsHardwareAccelerated && !Vector512.IsThrottling`, which seems "clear enough" for an internal only intrinsic.

Comment (DeepakRajendrakumaran) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:66]: Yeah - I think I prefer this too.

Comment (DeepakRajendrakumaran) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:66]: @tannergooding I'm having some trouble adding an internal property in Vector512. Outside of in Vector512.cs(https://github.com/dotnet/runtime/pull/86655/commits/49aa3168ef18de84b5d37ce47bd9c0400a3aa360#diff-97c3213a2a53c89fc5c8fd05ea68767d5b37e4fc5907825e814c150a55e3870dR55-R59), where else in library do I have to update? I added in JIT I think but I'm getting an 'UnSupported' exception
image

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:72]: You should just need to update it there and ensure it's handled in the JIT alongside the existing `IsHardwareAccelerated`/`IsSupported` handling.

I'd probably not make it recursive since its internal only. That avoids needing to ensure Mono handles it as well.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:72]: As in the method is currently defined as recursively calling itself:
```
internal static bool Vector512Throttling
{
    [Intrinsic]
    get => Vector512Throttling;
}
```

This is primarily important so that indirect invocation still works as expected and requires special JIT handling. That special JIT handling works implicitly for a `IsHardwareAccelerated` and `IsSupported`. New members, like `get_IsThrottled`, would need additional handling. Given that the API is internal, this support for indirect invocationisn't important and we can instead have it be `get => false;` which means it can only return `true` if called directly and if the runtime intrinsifies it and it means that Mono doesn't need to add explicit support for the new recursive member.

----------------------

An alternative is that we don't define this internal property at all. Instead, we'd have it surface indirectly via `IsHardwareAccelerated` itself, which allows external users to depend on it. That is, I think we have two viable paths here, each with their own tradeoffs. I don't have a strong preference for which we do in this PR. We probably want to discuss and determine what we think the best option before shipping is, however.

The first option is to have this internal `IsThrottled`/`IsThrottling` property and have the BCL use it as `Vector512.IsHardwareAccelerated && !Vector512.IsThrottled`. This allows the BCL to easily depend on it. However, it means users may end up needing to have their own equivalent to our internal `IsThrottled` property, replicating the checks we have internally.

The second option is to instead surface this directly through `Vector512.IsHardwareAccelerated`. That is, the JIT only returns `true` for `IsHardwareAccelerated` if `compSupportsHWIntrinsic(InstructionSet_AVX512F) && !jitFlags.IsSet(JitFlags::JIT_FLAG_VECTOR512_THROTTLING)`. The actual implementation would still use acceleration, much as `Vector256` does on hardware with AVX but without AVX2. The downside is that users on first generation hardware would have to explicitly check `Avx512F.IsSupported` if they wanted to light-up.

Its the same general support/consideration, just differing slightly in how its exposed and whether/how its visible to 3rd party consumers of the API.

Comment (DeepakRajendrakumaran) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:72]: I imagine we would eventually need to expose a 'IsThrottled' property anyway even for AVx512F for users who want to skip using AVX512 features when throttling is an issue. Considering that, I think option A(Exposing property) makes sense(maybe eventually making it public?).

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:72]: Either option gives that ability. It'd effectively be...

Option A:
* `Vector512.IsHardwareAccelerated` -- I don't care about throttling
* `Vector512.IsHardwareAccelerated && !Vector512.IsThrottled` -- I care about throttling

Option B:
* `Vector512.IsHardwareAccelerated` -- I care about throttling
* `Avx512F.IsSupported` -- I don't care about throttling

The main downside of `Option A` is that gives users "one more thing" to think about when presumably the common case is they want things to work just like `Vector128`/`Vector256`. It's also a bit quirky in that its exposing an xarch specific concern as part of the general API surface.

The main downside of `Option B` ends up being that users who are doing HPC or large workloads need to tie that functionality directly to an `Avx512F.IsSupported` check. 

It's something we should discuss a bit more in our next sync meeting. What we do for this PR, particularly while its internal, doesn't matter too much and we can always tweak it later.

Comment (DeepakRajendrakumaran) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:72]: Ah! Got it. Yeah- that makes sense.

Assuming this is going to be revisited later, I'm going to keep it to Option B. That way I'm not adding anything new to API surface which needs to be taken out. Adding something later seems better to me that taking it out. I'm okay with either though

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:72]: Works for me! Thanks for the discussion!

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs (hunk 6 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:74]: Certainly worth experimenting with internally at first.

File: src/coreclr/jit/hwintrinsicxarch.cpp (hunk 4 above)
Comment (DeepakRajendrakumaran) [reply src/coreclr/jit/hwintrinsicxarch.cpp]: Part of different PR

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs (hunk 7 above)
Comment (DeepakRajendrakumaran) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs]: Removed

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Packed.cs [hunk 8]
```diff
@@ -114,7 +114,7 @@ public static bool Contains(ref short searchSpace, short value, int length)
             {
                 ref short currentSearchSpace = ref searchSpace;
 #pragma warning disable IntrinsicsInSystemPrivateCoreLibAttributeNotSpecificEnough // The else condition for this if statement is identical in semantics to Avx2 specific code
-                if (Avx512BW.IsSupported && Vector512.IsHardwareAccelerated && length > Vector512<short>.Count)
+                if (Avx512F.IsSupported && Avx512BW.IsSupported && Vector512.IsHardwareAccelerated && length > Vector512<short>.Count)
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Packed.cs:117]: This should just need `Avx512BW.IsSupported` as it being true implies `Avx512F.IsSupported`. Same with other places below

We might even be able to simplify it down to `Vector512.IsHardwareAccelerated` + `Debug.Assert(Avx512BW.IsSupported)` given that the JIT currently requires `F + BW + CD + DQ + VL` and no other platforms support `V512`

File: src/coreclr/jit/compiler.h [hunk 9]
```diff
@@ -9474,6 +9474,16 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
             return jitFlags->IsSet(JitFlags::JIT_FLAG_REVERSE_PINVOKE);
         }
 
+        // true if JitFlags::JIT_FLAG_VECTOR512_THROTTLING is set to true
+        bool Vector512Throttling()
```
Comment (tannergooding) [inline src/coreclr/jit/compiler.h:9478]: nit: We should have the one other place `JIT_FLAG_VECTOR512_THROTTLING` is used in `compiler.cpp` call this method now.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs [hunk 10]
```diff
@@ -957,5 +968,19 @@ internal static uint FlipBit(uint value, int index)
...
+        /// Similar in behavior to the x86 instruction BTC (Bit Test and Complement).
+        /// /// </summary>
+        /// <param name="value">The value.</param>
+        /// <param name="index">The zero-based index of the bit to flip.
+        /// Any value outside the range [0..63] is treated as congruent mod 64.</param>
+        /// <returns>The new value.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static ulong FlipBit(ulong value, int index)
+        {
+            return value ^ (ulong)((ulong)1u << index);
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:983]:
```suggestion
            return value ^ (1UL << index);
```

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Packed.cs (hunk 8 above)
Comment (DeepakRajendrakumaran) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Packed.cs:117]: > We might even be able to simplify it down to `Vector512.IsHardwareAccelerated` + `Debug.Assert(Avx512BW.IsSupported)` given that the JIT currently requires `F + BW + CD + DQ + VL` and no other platforms support `V512`

I assumed the #pragma warning disable `IntrinsicsInSystemPrivateCoreLibAttributeNotSpecificEnough `and `compExactlyDependsOn` in called method means I need to have the `Avx512F.IsSupported `check.

File: src/coreclr/jit/hwintrinsic.cpp [hunk 11]
```diff
@@ -550,6 +550,12 @@ NamedIntrinsic HWIntrinsicInfo::lookupId(Compiler*         comp,
 
         if (isIsaSupported && comp->compSupportsHWIntrinsic(isa))
         {
+#ifdef TARGET_XARCH
+            if ((isa == InstructionSet_Vector512) && (comp->opts.Vector512Throttling()))
+            {
+                return NI_IsSupported_False;
+            }
```
[the remark this answers is not in the document]
Comment (DeepakRajendrakumaran) [reply src/coreclr/jit/hwintrinsic.cpp]: Done

File: src/coreclr/jit/compiler.h (hunk 9 above)
Comment (DeepakRajendrakumaran) [reply src/coreclr/jit/compiler.h:9478]: Done

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs (hunk 10 above)
Comment (DeepakRajendrakumaran) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs]: Done

**Comment by DeepakRajendrakumaran** [comment]
Ran Microbenchmark( `--filter "System.Memory.*"`) and compared with main branch : This was with following thresholds for ResultComparator(` --threshold 5% --noise 1ns`)



image

**Comment by DeepakRajendrakumaran** [comment]
@tannergooding Let me know if ther are any other changes you want me to do for this?

**Review by tannergooding — APPROVED** [review]

**Comment by tannergooding** [comment]
@stephentoub, @jeffhandley 

This should be ready for merge, just would be good to get a secondary sign-off. There's no perf diff for hardware without AVX-512 support since the entire code path is treated as dead code. Numbers generally look good for hardware with AVX-512 support with the additional branch causing some very minor TP hits for < 64 bytes of data.

The code is effectively the V256 code path, but duplicated. We can reduce this duplication longer term with the `ISimdVector<TSelf, TElement>` approach, but that should be its own PR.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs [hunk 12]
```diff
@@ -957,5 +968,19 @@ internal static uint FlipBit(uint value, int index)
...
+        /// Similar in behavior to the x86 instruction BTC (Bit Test and Complement).
+        /// /// </summary>
+        /// <param name="value">The value.</param>
+        /// <param name="index">The zero-based index of the bit to flip.
+        /// Any value outside the range [0..63] is treated as congruent mod 64.</param>
+        /// <returns>The new value.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static ulong FlipBit(ulong value, int index)
+        {
+            return value ^ (1UL << index);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:983]: Minor nit: we use `u` above but `UL` here... it'd be nice to be consistent on capitalization.

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Packed.cs [hunk 13]
```diff
@@ -113,9 +113,59 @@ public static bool Contains(ref short searchSpace, short value, int length)
             else
             {
                 ref short currentSearchSpace = ref searchSpace;
+#pragma warning disable IntrinsicsInSystemPrivateCoreLibAttributeNotSpecificEnough // The else condition for this if statement is identical in semantics to Avx2 specific code
+                if (Avx512BW.IsSupported && Vector512.IsHardwareAccelerated && length > Vector512<short>.Count)
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Packed.cs:117]: Why is the `Avx512BW.IsSupported` necessary?

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Packed.cs:117]: Nevermind, I see it's used by one of the helpers this calls.

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs [hunk 14]
```diff
@@ -3324,7 +3765,48 @@ internal static int LastIndexOfAnyExceptInRangeUnsignedNumber<T>(ref T searchSpa
...
+                        Vector256<T> targetVector256 = Vector256.Create(value);
+                        uint mask = Vector256.Equals(Vector256.LoadUnsafe(ref oneVectorAwayFromEnd), targetVector256).ExtractMostSignificantBits();
+
+                        // The mask contains some elements that may be double-checked, so shift them away in order to get the correct pop-count.
+                        uint overlaps = (uint)Vector256<T>.Count - remaining;
+                        mask >>= (int)overlaps;
+                        count += BitOperations.PopCount(mask);
+
+                        return count;
+                    }*/
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:3807]: What's the plan for this commented-out code and TODO?

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs (hunk 12 above)
Comment (DeepakRajendrakumaran) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs]: Changed to `ul`

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs (hunk 14 above)
Comment (DeepakRajendrakumaran) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs]: That was me experimenting with the microbenchmarks and checking what I see if I add some extra logic. I didn't see any improvement with it and that was meant to be removed. I have removed it now

**Comment by DeepakRajendrakumaran** [comment]
Bump. Anything else required from me here?

**Comment by BruceForstall** [comment]
@stephentoub @tannergooding Do you have any further requests or comments on this PR? E.g., any request for additional testing or performance analysis? Or is it ready to be merged?

**Review by stephentoub — APPROVED** [review]

**Comment by danmoseley** [comment]
Nice wins thanks @DeepakRajendrakumaran !
